docs/proposals/failover/application-failover.md

```diff
@@ -0,0 +1,506 @@
+    // be migrated along with the application.
+    // +optional
+    Application *ApplicationFailoverBehavior `json:"application,omitempty"`
+  
+    // Cluster indicates failover behaviors in case of cluster failure.
+    // If this value is nil, failover is disabled.
+    // +optional
+    Cluster *ClusterFailoverBehavior `json:"cluster,omitempty"`
+}
+
+// ApplicationFailoverBehavior indicates application failover behaviors.
+type ApplicationFailoverBehavior struct {
+    // InitiateGates indicates the preconditions of the failover process.
+    // If specified, all gates will be evaluated for prerequisites.
+    // +optional
+    InitiateGates []FailoverInitiateGate `json:"initiateGates,omitempty"`
+  
+    // ActiveGates indicates the conditions of performing the failover process.
+    //
+    // +kubebuilder:validation:MinItems=1
+    // +required
+    ActiveGates []FailoverActiveGate `json:"activeGates"`
+  
+    // TolerationSeconds represents the period of time the Karmada should wait
+    // before performing failover. Defaults to 10s.
+    //
+    // +kubebuilder:default=10
+    // +optional
+    TolerationSeconds int32
+  
+    // PurgeMode represents how to deal with the legacy applications on the
+    // cluster from which the application is migrated.
+    // Valid options are "Immediately", "Graciously" and "Never".
+    // Defaults to "Graciously".
+    //
+    // +kubebuilder:validation:Enum=Immediately;Graciously;Never
+    // +optional
+    PurgeMode string `json:"purgeMode,omitempty"`
+  
+    // EscapeSeconds represents the period of time the cluster from which the
+    // application was migrated from should be escaped.
+    // Defaults to 600s.
+    //
+    // +kubebuilder:default=600
+    // +optional
+    EscapeSeconds int32 `json:"escapeSeconds,omitempty"`
+}
+
+// FailoverInitiateGate represents the preconditions of the failover process.
+type FailoverInitiateGate struct {
+    // ConditionType refers to a condition in the object's condition list with
+    // matching type. If the status of the specified condition is "true" means pass.
+    // +optional
+    ConditionType string `json:"conditionType,omitempty"`
+  
+    // HealthyState refers to the healthy status reported by the Karmada resource
+    // interpreter.
+    // Valid options are "Healthy" and "Unknown".
+    // +optional
+    HealthyState string `json:"healthyState,omitempty"`
+}
+
+// FailoverActiveGate represents the conditions of performing the failover process.
+type FailoverActiveGate struct {
+    // ConditionType refers to a condition in the object's condition list with
+    // matching type. If the status of the specified condition is "true" means pass.
+    // +optional
+    ConditionType string `json:"conditionType,omitempty"`
+  
+    // HealthyState refers to the healthy status reported by the Karmada resource
+    // interpreter.
+    // Valid options are "Unhealthy" and "Unknown".
```

Comment:
Added two optional API designs. Meanwhile, I think `Unknown` may not be suitable as an option. As far as I know, it will only occur when users do not config the `InterpretHealth` or the config is wrong.